test/object-store/sync/flx_sync.cpp

```diff
@@ -3370,6 +3371,264 @@ TEST_CASE("flx: bootstrap changesets are applied continuously", "[sync][flx][app
+                    };
+                async_open->start(open_realm_completed_callback);
+                REQUIRE(open_realm_pf.future.get());
+                REQUIRE_FALSE(subscription_invoked.load());
+            }
+        }
+
+        SECTION("rerun on open set for multiple async open tasks (subscription runs only once)") {
+            auto subscription_pf = util::make_promise_future<bool>();
+            auto init_subscription_callback =
+                [&, promise_holder = util::CopyablePromiseHolder(std::move(subscription_pf.promise))](
+                    std::shared_ptr<Realm> realm) mutable {
+                    REQUIRE(realm);
+                    auto table = realm->read_group().get_table("class_TopLevel");
+                    Query query(table);
+                    auto subscription = realm->get_latest_subscription_set();
+                    auto mutable_subscription = subscription.make_mutable_copy();
+                    mutable_subscription.insert_or_assign(query);
+                    mutable_subscription.commit();
+                    promise_holder.get_promise().emplace_value(true);
+                };
+            config.sync_config->subscription_initializer = init_subscription_callback;
+            config.sync_config->rerun_init_subscription_on_open = true;
+
+            auto async_open_task1 = Realm::get_synchronized_realm(config);
+            auto async_open_task2 = Realm::get_synchronized_realm(config);
+
+            auto open_t1_pf = util::make_promise_future<bool>();
+            auto open_t2_pf = util::make_promise_future<bool>();
+
+            auto open_callback_task_1 = [&,
+                                         promise_holder = util::CopyablePromiseHolder(std::move(open_t1_pf.promise))](
+                                            ThreadSafeReference ref, std::exception_ptr err) mutable {
+                REQUIRE_FALSE(err);
+                auto realm = Realm::get_shared_realm(std::move(ref));
+                REQUIRE(realm);
+                promise_holder.get_promise().emplace_value(true);
+            };
+            auto open_callback_task_2 = [&,
+                                         promise_holder = util::CopyablePromiseHolder(std::move(open_t2_pf.promise))](
+                                            ThreadSafeReference ref, std::exception_ptr err) mutable {
+                REQUIRE_FALSE(err);
+                auto realm = Realm::get_shared_realm(std::move(ref));
+                REQUIRE(realm);
+                promise_holder.get_promise().emplace_value(true);
+            };
+
+            async_open_task1->start(open_callback_task_1);
+            async_open_task2->start(open_callback_task_2);
+
+            // subscription init called only once but realm opened twice
+            REQUIRE(open_t1_pf.future.get());
+            REQUIRE(open_t2_pf.future.get());
+            REQUIRE(subscription_pf.future.get());
```

Comment:
We should check only one subscription was committed:
```
REQUIRE(realm->get_latest_subscription_set().version() == 1);
REQUIRE(realm->get_active_subscription_set().version() == 1);
```
We need to get the realm instance through one of the futures used in the complete callbacks.